File: src/term/grid.rs
use crate::term::color::Color;
use bitflags::bitflags;
use std::collections::VecDeque;

bitflags! {
    #[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
    pub struct CellFlags: u8 {
        const BOLD = 1 << 0;
        const ITALIC = 1 << 1;
        const UNDERLINE = 1 << 2;
        const REVERSE = 1 << 3;
        /// First column of a double-width glyph.
        const WIDE = 1 << 4;
        /// Trailing placeholder column following a WIDE cell.
        const WIDE_SPACER = 1 << 5;
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Cell {
    pub c: char,
    pub fg: Color,
    pub bg: Color,
    pub flags: CellFlags,
}

impl Default for Cell {
    fn default() -> Self {
        Cell {
            c: ' ',
            fg: Color::Default,
            bg: Color::Default,
            flags: CellFlags::empty(),
        }
    }
}

pub type Row = Vec<Cell>;

pub struct Grid {
    pub cols: usize,
    pub rows: usize,
    lines: Vec<Row>,
    /// Parallel to `lines`: `wrapped[i]` means row `i`'s content is a
    /// terminal-forced continuation of the same logical line as row
    /// `i + 1` (as opposed to row `i + 1` starting a fresh line because of
    /// an actual `\n`). Used to reflow content across a column-count
    /// resize instead of truncating it. Always kept the same length as
    /// `lines`.
    wrapped: Vec<bool>,
    pub scrollback: VecDeque<Row>,
    pub scrollback_limit: usize,
    /// How many rows at the back of `scrollback` were hidden by a resize
    /// shrink a moment ago and are still safe for a subsequent grow to
    /// pull back undisturbed. Paired with `scrollback_len_when_pending`,
    /// which pins down the exact `scrollback.len()` those rows were
    /// pushed at: if anything else has touched `scrollback` since (real
    /// output scrolling off via `scroll_up`, or the cap evicting from the
    /// front), the length no longer matches and the pending rows are
    /// treated as gone rather than popped. Without this a grow following
    /// a shrink would otherwise blindly pop whatever is currently at the
    /// back of `scrollback` -- which, if the shell printed anything in
    /// between, is real output that already scrolled off for good, not
    /// this resize's own hidden row -- resurrecting already-displayed
    /// text at the top of the pane.
    rows_pending_restore: usize,
    scrollback_len_when_pending: usize,
}

impl Grid {
    pub fn new(cols: usize, rows: usize, scrollback_limit: usize) -> Self {
        Grid {
            cols,
            rows,
            lines: vec![vec![Cell::default(); cols]; rows],
            wrapped: vec![false; rows],
            scrollback: VecDeque::new(),
            scrollback_limit,
            rows_pending_restore: 0,
            scrollback_len_when_pending: 0,
        }
    }

    fn blank_row(&self) -> Row {
        vec![Cell::default(); self.cols]
    }

    pub fn row_mut(&mut self, idx: usize) -> &mut Row {
        &mut self.lines[idx]
    }

    // Only exercised by tests today (production code reads whole rows via
    // `line_at`), so it's flagged dead code outside `cargo test`.
    #[cfg_attr(not(test), allow(dead_code))]
    pub fn cell(&self, row: usize, col: usize) -> &Cell {
        &self.lines[row][col]
    }

    pub fn cell_mut(&mut self, row: usize, col: usize) -> &mut Cell {
        &mut self.lines[row][col]
    }

    /// Mark whether `row`'s content continues onto the next row because a
    /// print forced a line wrap there (as opposed to an actual `\n`).
    /// Called by `Term::print_char` at the moment a wrap actually happens.
    pub fn set_wrapped(&mut self, row: usize, wrapped: bool) {
        if let Some(slot) = self.wrapped.get_mut(row) {
            *slot = wrapped;
        }
    }

    /// Resolves a viewport row to its source line, accounting for how far
    /// back `scroll_offset` has scrolled into scrollback. `view_row` is
    /// clamped to `0..rows` by the caller.
    pub fn line_at(&self, view_row: usize, scroll_offset: usize) -> &Row {
        let scroll_offset = scroll_offset.min(self.scrollback.len());
        if view_row < scroll_offset {
            &self.scrollback[self.scrollback.len() - scroll_offset + view_row]
        } else {
            &self.lines[view_row - scroll_offset]
        }
    }

    /// A viewport row's distance from the bottom of the live screen, in
    /// lines -- 0 is the bottom-most live row, `rows - 1` the top-most
    /// live row, `rows` the most recently scrolled-off line, and so on
    /// back through scrollback. Selection endpoints are stored in this
    /// form (see `crate::tab::Selection`) instead of raw `(view_row,
    /// scroll_offset)` pairs specifically so a selection stays anchored to
    /// the same text while the user scrolls mid-drag: `view_row` alone is
    /// meaningless once `scroll_offset` changes, but a line's distance
    /// from the live bottom does not.
    pub fn distance_from_bottom(&self, view_row: usize, scroll_offset: usize) -> usize {
        scroll_offset + (self.rows - 1 - view_row)
    }

    /// The inverse of `distance_from_bottom`: resolves a stored selection
    /// endpoint back to the row it refers to, or `None` if that line has
    /// since fallen out of scrollback (e.g. the scrollback cap shrank).
    pub fn absolute_line(&self, distance_from_bottom: usize) -> Option<&Row> {
        if distance_from_bottom < self.rows {
            Some(&self.lines[self.rows - 1 - distance_from_bottom])
        } else {
            let k = distance_from_bottom - self.rows;
            (k < self.scrollback.len()).then(|| &self.scrollback[self.scrollback.len() - 1 - k])
        }
    }

    /// Resize to a new column/row count, reflowing content instead of
    /// destroying it: rows linked by `wrapped` are treated as one logical
    /// line, re-wrapped at the new column count, so text that no longer
    /// fits on one row moves to the next instead of being truncated (and
    /// comes back correctly if the terminal widens again). Rows that
    /// overflow the new row count are pushed into scrollback (or pulled
    /// back out of it, if growing) rather than dropped.
    ///
    /// `cursor` is the pre-resize (row, col); returns where it lands
    /// after reflowing.
    pub fn resize_reflow(&mut self, cols: usize, rows: usize, cursor: (usize, usize)) -> (usize, usize) {
        if cols == self.cols && rows == self.rows {
            return cursor;
        }
        let (cursor_row, cursor_col) = if cols != self.cols {
            self.reflow_columns(cols, cursor)
        } else {
            cursor
        };
        let cursor_row = self.adjust_row_count(rows, cursor_row);
        self.cols = cols;
        self.rows = rows;
        (cursor_row.min(rows - 1), cursor_col.min(cols.saturating_sub(1)))
    }

    /// Resize to a new column/row count without reflowing: rows are
    /// truncated or padded with blanks, columns are truncated or padded
    /// per row. Used for the alternate screen, since full-screen apps
    /// (vim, htop, ...) redraw themselves completely on `SIGWINCH` rather
    /// than relying on the terminal to preserve their content.
    pub fn resize_truncate(&mut self, cols: usize, rows: usize) {
        if cols == self.cols && rows == self.rows {
            return;
        }
        for row in &mut self.lines {
            row.resize(cols, Cell::default());
        }
        self.lines.resize(rows, vec![Cell::default(); cols]);
        self.wrapped.resize(rows, false);
        self.cols = cols;
        self.rows = rows;
    }

    fn reflow_columns(&mut self, new_cols: usize, cursor: (usize, usize)) -> (usize, usize) {
        let old_cols = self.cols;
        let cursor_row = cursor.0.min(self.lines.len().saturating_sub(1));
        let cursor_col = cursor.1.min(old_cols.saturating_sub(1));

        // Group rows into logical lines (runs linked by `wrapped`),
        // flattening each into one cell sequence, and remember where the
        // cursor falls in that flattened representation.
        let mut logical_lines: Vec<Vec<Cell>> = Vec::new();
        let mut cursor_logical = 0usize;
        let mut cursor_offset = 0usize;

        let mut row_idx = 0;
        while row_idx < self.lines.len() {
            let mut cells: Vec<Cell> = Vec::new();
            let is_cursor_line_start = logical_lines.len();
            loop {
                if row_idx == cursor_row {
                    cursor_logical = is_cursor_line_start;
                    cursor_offset = cells.len() + cursor_col;
                }
                cells.extend_from_slice(&self.lines[row_idx]);
                let wraps = self.wrapped.get(row_idx).copied().unwrap_or(false);
                row_idx += 1;
                if !wraps || row_idx >= self.lines.len() {
                    break;
                }
            }
            // Trim trailing blank cells -- only meaningful at the very end
            // of the logical line, since interior rows are always full
            // (that's why they wrapped). A fully blank line trims to
            // empty; the re-wrap loop below turns that back into exactly
            // one blank row.
            while cells.last() == Some(&Cell::default()) {
                cells.pop();
            }
            if logical_lines.len() == cursor_logical {
                cursor_offset = cursor_offset.min(cells.len());
            }
            logical_lines.push(cells);
        }

        // Drop trailing blank logical lines (almost always most of the
        // unused bottom of the screen) rather than re-wrapping each into
        // its own padding row -- that would keep inflating the row count
        // every time a narrower resize adds more wrapped rows, pushing
        // real content into scrollback for no reason. The one holding the
        // cursor is kept even if blank, so it still has somewhere to land.
        while let Some(last) = logical_lines.last() {
            let is_cursor_line = logical_lines.len() - 1 == cursor_logical;
            if last.is_empty() && !is_cursor_line {
                logical_lines.pop();
            } else {
                break;
            }
        }
        if logical_lines.is_empty() {
            logical_lines.push(Vec::new());
            cursor_logical = 0;
            cursor_offset = 0;
        }

        // Re-wrap each logical line at `new_cols`.
        let mut new_lines: Vec<Row> = Vec::new();
        let mut new_wrapped: Vec<bool> = Vec::new();
        let mut new_cursor = (0usize, 0usize);

        for (logical_idx, cells) in logical_lines.into_iter().enumerate() {
            let is_cursor_line = logical_idx == cursor_logical;
            let mut offset = 0usize;
            loop {
                let end = (offset + new_cols).min(cells.len());
                let mut row: Row = cells[offset..end].to_vec();
                row.resize(new_cols, Cell::default());

                if is_cursor_line && cursor_offset >= offset && cursor_offset < offset + new_cols {
                    new_cursor = (new_lines.len(), cursor_offset - offset);
                }

                let more = end < cells.len();
                new_wrapped.push(more);
                new_lines.push(row);

                if !more {
                    if is_cursor_line && cursor_offset >= end {
                        new_cursor = (new_lines.len() - 1, (cursor_offset - offset).min(new_cols.saturating_sub(1)));
                    }
                    break;
                }
                offset += new_cols;
            }
        }

        if new_lines.is_empty() {
            new_lines.push(vec![Cell::default(); new_cols]);
            new_wrapped.push(false);
        }

        self.lines = new_lines;
        self.wrapped = new_wrapped;
        self.cols = new_cols;
        new_cursor
    }

    /// Grow or shrink the row count by pushing/pulling rows through
    /// scrollback (like normal scrolling), instead of truncating. Rows are
    /// only ever added/removed at the top (where scrollback connects), so
    /// takes and returns the cursor's row index, shifted to stay attached
    /// to the same row through any insertions/removals there.
    fn adjust_row_count(&mut self, rows: usize, mut cursor_row: usize) -> usize {
        while self.lines.len() > rows {
            let removed = self.lines.remove(0);
            self.wrapped.remove(0);
            self.scrollback.push_back(removed);
            if self.scrollback.len() > self.scrollback_limit {
                self.scrollback.pop_front();
            }
            self.rows_pending_restore += 1;
            self.scrollback_len_when_pending = self.scrollback.len();
            cursor_row = cursor_row.saturating_sub(1);
        }
        while self.lines.len() < rows {
            // Scrollback isn't reflowed when the column count changes, so
            // a row stored at an old width can't just be spliced back in
            // here -- that would leave `lines` with rows of mismatched
            // length, corrupting every column-indexed access after it.
            let matches_width = self.scrollback.back().is_some_and(|r| r.len() == self.cols);
            // `scrollback.len()` must still be exactly what it was right
            // after the pending rows were pushed -- if it's higher (real
            // output scrolled off in between) or lower (the cap evicted
            // from the front), something else has touched `scrollback`
            // since, and the tail is no longer guaranteed to be ours.
            let pending_intact = self.rows_pending_restore > 0 && self.scrollback.len() == self.scrollback_len_when_pending;
            if pending_intact && matches_width {
                let row = self.scrollback.pop_back().expect("checked Some above");
                self.lines.insert(0, row);
                self.wrapped.insert(0, false);
                self.rows_pending_restore -= 1;
                self.scrollback_len_when_pending -= 1;
                cursor_row += 1;
            } else {
                self.rows_pending_restore = 0;
                // Padding with a blank row at the bottom doesn't shift any
                // existing row's index.
                self.lines.push(self.blank_row());
                self.wrapped.push(false);
            }
        }
        cursor_row
    }

    /// Update the scrollback cap, trimming immediately if it shrank. Used
    /// when settings are changed live from the settings window.
    pub fn set_scrollback_limit(&mut self, limit: usize) {
        self.scrollback_limit = limit;
        while self.scrollback.len() > self.scrollback_limit {
            self.scrollback.pop_front();
        }
    }

    /// Scroll the inclusive region [top, bottom] up by `n` lines, filling
    /// the vacated bottom lines with blanks. Lines scrolled off the top are
    /// only kept in scrollback when the region spans the whole screen (a
    /// scroll-region-limited scroll, e.g. inside vim, must not pollute
    /// scrollback with lines that never really left the visible screen).
    pub fn scroll_up(&mut self, top: usize, bottom: usize, n: usize) {
        let region_len = bottom + 1 - top;
        let n = n.min(region_len);
        let push_scrollback = top == 0 && bottom == self.rows - 1;
        for _ in 0..n {
            let removed = self.lines.remove(top);
            self.wrapped.remove(top);
            if push_scrollback {
                self.scrollback.push_back(removed);
                if self.scrollback.len() > self.scrollback_limit {
                    self.scrollback.pop_front();
                }
            }
            self.lines.insert(bottom, self.blank_row());
            self.wrapped.insert(bottom, false);
        }
    }

    /// Scroll the inclusive region [top, bottom] down by `n` lines (reverse
    /// index), filling the vacated top lines with blanks.
    pub fn scroll_down(&mut self, top: usize, bottom: usize, n: usize) {
        let region_len = bottom + 1 - top;
        let n = n.min(region_len);
        for _ in 0..n {
            self.lines.remove(bottom);
            self.wrapped.remove(bottom);
            self.lines.insert(top, self.blank_row());
            self.wrapped.insert(top, false);
        }
    }

    pub fn clear_all(&mut self) {
        for row in &mut self.lines {
            row.fill(Cell::default());
        }
        self.wrapped.fill(false);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn distance_from_bottom_spans_live_rows_at_zero_scroll() {
        let grid = Grid::new(10, 4, 100);
        // Bottom row is 0 lines away from itself; the top row is 3 (rows
        // - 1) lines above it.
        assert_eq!(grid.distance_from_bottom(3, 0), 0);
        assert_eq!(grid.distance_from_bottom(0, 0), 3);
    }

    #[test]
    fn distance_from_bottom_accounts_for_scroll_offset() {
        let grid = Grid::new(10, 4, 100);
        // Scrolled up by 2: the viewport's bottom row is now showing what
        // used to be 2 lines above the true bottom.
        assert_eq!(grid.distance_from_bottom(3, 2), 2);
        assert_eq!(grid.distance_from_bottom(0, 2), 5);
    }

    #[test]
    fn absolute_line_is_the_inverse_of_distance_from_bottom() {
        let mut grid = Grid::new(3, 2, 100);
        grid.row_mut(0)[0].c = 'A'; // top live row
        grid.row_mut(1)[0].c = 'B'; // bottom live row
        assert_eq!(grid.absolute_line(0).unwrap()[0].c, 'B');
        assert_eq!(grid.absolute_line(1).unwrap()[0].c, 'A');
    }

    #[test]
    fn absolute_line_reaches_into_scrollback() {
        let mut grid = Grid::new(3, 2, 100);
        let mut pushed = grid.blank_row();
        pushed[0].c = 'S';
        grid.scrollback.push_back(pushed);
        // Past the two live rows (distance 0, 1), distance 2 is the most
        // recently scrolled-off line.
        assert_eq!(grid.absolute_line(2).unwrap()[0].c, 'S');
    }

    #[test]
    fn absolute_line_none_past_scrollback_end() {
        let grid = Grid::new(3, 2, 100);
        assert!(grid.absolute_line(2).is_none());
    }
}

File: swift/Sources/KetermCore/Grid.swift
/// The character grid: the live screen plus, for the primary screen, the
/// scrollback behind it.
///
/// Rows are addressed two ways, deliberately. `row(_:)` indexes the live
/// screen top-down, which is what the parser works in. Selections and
/// search results instead store a *distance from the bottom* (see
/// `distanceFromBottom`), because a screen row number stops meaning
/// anything the moment the view scrolls, while "three lines above the
/// live bottom" keeps pointing at the same text.
public struct Grid: Sendable {
    public private(set) var columns: Int
    public private(set) var rows: Int
    private var lines: [[Cell]]
    /// Parallel to `lines`: `wrapped[i]` means row `i` runs on into row
    /// `i + 1` because printing hit the right margin, as opposed to the
    /// program having sent a newline. Reflowing a column-count change
    /// needs this to know which rows are one logical line.
    private var wrapped: [Bool]
    public private(set) var scrollback: [[Cell]] = []
    public var scrollbackLimit: Int

    /// How many rows at the end of `scrollback` were hidden by a resize
    /// shrink and are still safe for a later grow to pull back, paired
    /// with the exact `scrollback.count` they were pushed at.
    ///
    /// Both are needed: without the count, a grow after the shell printed
    /// something in between would pop that new output back onto the
    /// screen instead of the row the shrink hid -- text already visible
    /// elsewhere reappearing at the top. Found the hard way in the Rust
    /// version; the pairing is what makes "is the tail still ours?"
    /// answerable.
    private var rowsPendingRestore = 0
    private var scrollbackCountWhenPending = 0

    public init(columns: Int, rows: Int, scrollbackLimit: Int) {
        self.columns = max(1, columns)
        self.rows = max(1, rows)
        self.scrollbackLimit = scrollbackLimit
        self.lines = Array(repeating: Array(repeating: Cell.blank, count: self.columns), count: self.rows)
        self.wrapped = Array(repeating: false, count: self.rows)
    }

    private var blankRow: [Cell] { Array(repeating: Cell.blank, count: columns) }

    public func row(_ index: Int) -> [Cell] { lines[index] }

    public subscript(row: Int, column: Int) -> Cell {
        get { lines[row][column] }
        set { lines[row][column] = newValue }
    }

    /// Mark whether printing wrapped off the end of `row` onto the next.
    public mutating func setWrapped(_ row: Int, _ value: Bool) {
        guard lines.indices.contains(row) else { return }
        wrapped[row] = value
    }

    /// Resolve a viewport row to its source line, accounting for how far
    /// `scrollOffset` has scrolled back. `viewRow` is clamped to
    /// `0..<rows` by the caller.
    public func line(viewRow: Int, scrollOffset: Int) -> [Cell] {
        let offset = min(scrollOffset, scrollback.count)
        if viewRow < offset {
            return scrollback[scrollback.count - offset + viewRow]
        }
        return lines[viewRow - offset]
    }

    /// A viewport row's distance from the live bottom: 0 is the bottom
    /// row, `rows - 1` the top one, `rows` the most recently scrolled-off
    /// line, and so on back through scrollback.
    public func distanceFromBottom(viewRow: Int, scrollOffset: Int) -> Int {
        scrollOffset + (rows - 1 - viewRow)
    }

    /// The inverse of `distanceFromBottom`, or `nil` once that line has
    /// fallen out of scrollback.
    public func absoluteLine(distanceFromBottom distance: Int) -> [Cell]? {
        if distance < rows {
            return lines[rows - 1 - distance]
        }
        let k = distance - rows
        guard k < scrollback.count else { return nil }
        return scrollback[scrollback.count - 1 - k]
    }

    /// Total addressable lines: the live screen plus scrollback.
    public var totalLines: Int { rows + scrollback.count }

    public mutating func clearAll() {
        for i in lines.indices { lines[i] = blankRow }
        for i in wrapped.indices { wrapped[i] = false }
    }

    public mutating func clearScrollback() {
        scrollback.removeAll()
        rowsPendingRestore = 0
    }

    public mutating func setScrollbackLimit(_ limit: Int) {
        scrollbackLimit = limit
        while scrollback.count > scrollbackLimit { scrollback.removeFirst() }
    }

    // MARK: - Scrolling

    /// Scroll `[top, bottom]` up by `n`, blanking the vacated rows.
    /// Lines leaving the top are kept in scrollback only when the region
    /// spans the whole screen: a scroll confined to a region (what vim
    /// does inside a split) never really left the screen, and pushing it
    /// would pollute the history with text the user can still see.
    public mutating func scrollUp(top: Int, bottom: Int, count n: Int) {
        rotateUp(top: top, bottom: bottom, count: n, keepingHistory: top == 0 && bottom == rows - 1)
    }

    /// Delete lines (DL): the same rotation, but the removed lines are
    /// discarded. An app deleting lines out of the middle of the screen
    /// isn't scrolling them off, and they must not turn up in scrollback.
    public mutating func deleteLines(top: Int, bottom: Int, count n: Int) {
        rotateUp(top: top, bottom: bottom, count: n, keepingHistory: false)
    }

    private mutating func rotateUp(top: Int, bottom: Int, count n: Int, keepingHistory: Bool) {
        let regionLength = bottom + 1 - top
        let n = min(n, regionLength)
        guard n > 0 else { return }
        for _ in 0..<n {
            let removed = lines.remove(at: top)
            wrapped.remove(at: top)
            if keepingHistory {
                scrollback.append(removed)
                if scrollback.count > scrollbackLimit { scrollback.removeFirst() }
            }
            lines.insert(blankRow, at: bottom)
            wrapped.insert(false, at: bottom)
        }
    }

    /// Scroll `[top, bottom]` down by `n` (reverse index), blanking the
    /// vacated rows at the top.
    public mutating func scrollDown(top: Int, bottom: Int, count n: Int) {
        let regionLength = bottom + 1 - top
        let n = min(n, regionLength)
        guard n > 0 else { return }
        for _ in 0..<n {
            lines.remove(at: bottom)
            wrapped.remove(at: bottom)
            lines.insert(blankRow, at: top)
            wrapped.insert(false, at: top)
        }
    }

    // MARK: - Resize

    /// Resize without reflowing: rows and columns are truncated or padded
    /// with blanks. Used for the alternate screen, where full-screen apps
    /// redraw themselves on `SIGWINCH` instead of expecting the terminal
    /// to preserve anything.
    public mutating func resizeTruncating(columns newColumns: Int, rows newRows: Int) {
        let newColumns = max(1, newColumns), newRows = max(1, newRows)
        guard newColumns != columns || newRows != rows else { return }
        for i in lines.indices {
            if lines[i].count > newColumns {
                lines[i].removeLast(lines[i].count - newColumns)
            } else {
                lines[i].append(contentsOf: Array(repeating: Cell.blank, count: newColumns - lines[i].count))
            }
        }
        columns = newColumns
        if lines.count > newRows {
            lines.removeLast(lines.count - newRows)
            wrapped.removeLast(wrapped.count - newRows)
        } else {
            lines.append(contentsOf: Array(repeating: blankRow, count: newRows - lines.count))
            wrapped.append(contentsOf: Array(repeating: false, count: newRows - wrapped.count))
        }
        rows = newRows
    }

    /// Resize the primary screen, reflowing rather than destroying:
    /// rows linked by `wrapped` are one logical line, re-wrapped at the
    /// new width, so narrowing the window moves text onto more rows
    /// instead of cutting it off -- and widening puts it back. Rows that
    /// no longer fit move through scrollback rather than being dropped.
    ///
    /// Takes and returns the cursor's `(row, column)` so it stays on the
    /// same character across the reflow.
    public mutating func resizeReflowing(columns newColumns: Int, rows newRows: Int, cursor: (row: Int, column: Int)) -> (row: Int, column: Int) {
        let newColumns = max(1, newColumns), newRows = max(1, newRows)
        guard newColumns != columns || newRows != rows else { return cursor }

        var placed = cursor
        if newColumns != columns {
            placed = reflowColumns(to: newColumns, cursor: cursor)
        }
        placed.row = adjustRowCount(to: newRows, cursorRow: placed.row)
        columns = newColumns
        rows = newRows
        return (min(placed.row, newRows - 1), min(placed.column, newColumns - 1))
    }

    private mutating func reflowColumns(to newColumns: Int, cursor: (row: Int, column: Int)) -> (row: Int, column: Int) {
        let cursorRow = min(cursor.row, max(0, lines.count - 1))
        let cursorColumn = min(cursor.column, max(0, columns - 1))

        // Gather rows into logical lines, remembering where the cursor
        // falls within the flattened run.
        var logical: [[Cell]] = []
        var cursorLogical = 0
        var cursorOffset = 0
        var index = 0
        while index < lines.count {
            var cells: [Cell] = []
            let lineStart = logical.count
            while true {
                if index == cursorRow {
                    cursorLogical = lineStart
                    cursorOffset = cells.count + cursorColumn
                }
                cells.append(contentsOf: lines[index])
                let continues = index < wrapped.count && wrapped[index]
                index += 1
                if !continues || index >= lines.count { break }
            }
            // Trailing blanks are only meaningful at the very end of a
            // logical line -- interior rows are full by definition, which
            // is why they wrapped.
            while cells.last == Cell.blank { cells.removeLast() }
            if logical.count == cursorLogical { cursorOffset = min(cursorOffset, cells.count) }
            logical.append(cells)
        }

        // Drop trailing blank logical lines instead of re-wrapping each
        // into a padding row: keeping them would inflate the row count
        // every time a narrower resize adds wrapped rows, pushing real
        // content into scrollback for nothing. The one holding the cursor
        // stays, so it has somewhere to land.
        while let last = logical.last, last.isEmpty, logical.count - 1 != cursorLogical {
            logical.removeLast()
        }
        if logical.isEmpty {
            logical = [[]]
            cursorLogical = 0
            cursorOffset = 0
        }

        var rebuilt: [[Cell]] = []
        var rebuiltWrapped: [Bool] = []
        var newCursor = (row: 0, column: 0)

        for (logicalIndex, cells) in logical.enumerated() {
            let isCursorLine = logicalIndex == cursorLogical
            var offset = 0
            while true {
                let end = min(offset + newColumns, cells.count)
                var row = Array(cells[offset..<end])
                row.append(contentsOf: Array(repeating: Cell.blank, count: newColumns - row.count))

                if isCursorLine, cursorOffset >= offset, cursorOffset < offset + newColumns {
                    newCursor = (rebuilt.count, cursorOffset - offset)
                }
                let more = end < cells.count
                rebuiltWrapped.append(more)
                rebuilt.append(row)
                if !more {
                    if isCursorLine, cursorOffset >= end {
                        newCursor = (rebuilt.count - 1, min(cursorOffset - offset, newColumns - 1))
                    }
                    break
                }
                offset += newColumns
            }
        }
        if rebuilt.isEmpty {
            rebuilt = [Array(repeating: Cell.blank, count: newColumns)]
            rebuiltWrapped = [false]
        }

        lines = rebuilt
        wrapped = rebuiltWrapped
        columns = newColumns
        return newCursor
    }

    /// Grow or shrink the row count by moving rows through scrollback
    /// rather than truncating. Rows only ever leave or arrive at the top,
    /// where scrollback connects, so the cursor row shifts with them.
    private mutating func adjustRowCount(to newRows: Int, cursorRow: Int) -> Int {
        var cursorRow = cursorRow
        while lines.count > newRows {
            let removed = lines.removeFirst()
            wrapped.removeFirst()
            scrollback.append(removed)
            if scrollback.count > scrollbackLimit { scrollback.removeFirst() }
            rowsPendingRestore += 1
            scrollbackCountWhenPending = scrollback.count
            cursorRow = max(0, cursorRow - 1)
        }
        while lines.count < newRows {
            // Two conditions, both required. The width must still match:
            // scrollback isn't reflowed on a column change, and splicing
            // a differently-sized row back in would corrupt every
            // column-indexed access after it. And the tail must still be
            // what this shrink pushed -- see `rowsPendingRestore`.
            let widthMatches = scrollback.last?.count == columns
            let pendingIntact = rowsPendingRestore > 0 && scrollback.count == scrollbackCountWhenPending
            if widthMatches && pendingIntact {
                lines.insert(scrollback.removeLast(), at: 0)
                wrapped.insert(false, at: 0)
                rowsPendingRestore -= 1
                scrollbackCountWhenPending -= 1
                cursorRow += 1
            } else {
                rowsPendingRestore = 0
                // Padding at the bottom doesn't shift any row's index.
                lines.append(blankRow)
                wrapped.append(false)
            }
        }
        return cursorRow
    }
}
